packages/edit-navigation/src/store/actions.js

```diff
@@ -0,0 +1,251 @@
+			'core/notices',
+			'createErrorNotice',
+			__( 'There was an error.' ),
+			{
+				type: 'snackbar',
+			}
+		);
+	}
+} );
+
+function mapMenuItemsByClientId( menuItems, clientIdsByMenuId ) {
+	const result = {};
+	if ( ! menuItems || ! clientIdsByMenuId ) {
+		return result;
+	}
+	for ( const menuItem of menuItems ) {
+		const clientId = clientIdsByMenuId[ menuItem.id ];
+		if ( clientId ) {
+			result[ clientId ] = menuItem;
+		}
+	}
+	return result;
+}
+
+function* batchSave( menuId, menuItemsByClientId, navigationBlock ) {
+	const { nonce, stylesheet } = yield apiFetch( {
+		path: '/__experimental/customizer-nonces/get-save-nonce',
+	} );
+
+	// eslint-disable-next-line no-undef
+	const body = new FormData();
+	body.append( 'wp_customize', 'on' );
+	body.append( 'customize_theme', stylesheet );
+	body.append( 'nonce', nonce );
+	body.append( 'customize_changeset_uuid', uuidv4() );
+	body.append( 'customize_autosaved', 'on' );
+	body.append( 'customize_changeset_status', 'publish' );
+	body.append( 'action', 'customize_save' );
+	body.append(
+		'customized',
+		computeCustomizedAttribute(
+			navigationBlock.innerBlocks,
+			menuId,
+			menuItemsByClientId
+		)
+	);
+
+	yield apiFetch( {
+		url: '/wp-admin/admin-ajax.php',
+		method: 'POST',
+		body,
+	} );
+}
+
+function computeCustomizedAttribute( blocks, menuId, menuItemsByClientId ) {
+	const blocksList = blocksTreeToFlatList( blocks );
+	const dataList = blocksList.map( ( { block, parentId, position } ) =>
+		linkBlockToRequestItem( block, parentId, position )
+	);
+
+	// Create an object like { "nav_menu_item[12]": {...}} }
+	const computeKey = ( item ) => `nav_menu_item[${ item.id }]`;
+	const dataObject = keyBy( dataList, computeKey );
+
+	// Deleted menu items should be sent as false, e.g. { "nav_menu_item[13]": false }
+	for ( const clientId in menuItemsByClientId ) {
+		const key = computeKey( menuItemsByClientId[ clientId ] );
+		if ( ! ( key in dataObject ) ) {
+			dataObject[ key ] = false;
+		}
+	}
+
+	return JSON.stringify( dataObject );
+
+	function blocksTreeToFlatList( innerBlocks, parentId = 0 ) {
+		return innerBlocks.flatMap( ( block, index ) =>
+			[ { block, parentId, position: index + 1 } ].concat(
+				blocksTreeToFlatList(
+					block.innerBlocks,
+					getMenuItemForBlock( block )?.id
+				)
+			)
+		);
+	}
+
+	function linkBlockToRequestItem( block, parentId, position ) {
+		const menuItem = omit( getMenuItemForBlock( block ), 'menus', 'meta' );
+		return {
+			...menuItem,
+			position,
+			title: block.attributes?.label,
+			url: block.attributes.url,
+			original_title: '',
+			classes: ( menuItem.classes || [] ).join( ' ' ),
+			xfn: ( menuItem.xfn || [] ).join( ' ' ),
+			nav_menu_term_id: menuId,
+			menu_item_parent: parentId,
+			status: 'publish',
+			_invalid: false,
+		};
+	}
+
+	function getMenuItemForBlock( block ) {
+		return omit( menuItemsByClientId[ block.clientId ] || {}, '_links' );
+	}
+}
+
+function serializeProcessing( callback ) {
```

Comment:
This wrapper guarantees serial execution of data processing actions. `saveNavigationPost()` needs to wait for all the missing items to be created, and running multiple `createMissingMenuItems()` concurrently could result in sending more requests than required.